### ISSUE TITLE
setup>functions>conditional measurement stops working all the time

### ISSUE THREAD

**Issue by SAM26K** [kizniche/Mycodo#571]
## Mycodo Issue Report:
  - Specific Mycodo Version: 6.4.5
 
#### Problem Description
Setup conditional measurement control to turn on a relay.  Conditional control stops working after some time.  Works for hours sometimes.  Only way to fix is change some parameters in the conditional measurement control and save the changes.

  - what were you trying to do
  using analog sensors to turn a relay on and off depending on the voltage from the sensors.

**Comment by kizniche** [comment]
Does the daemon log contain errors?

Please screenshot both the conditional and input settings.

**Comment by SAM26K** [comment]
Hi Kyle,
Sorry for the delay.  When it happens the controls still switchd the relays on and off manually. Here's the screenshots:
![input1](https://user-images.githubusercontent.com/22091149/49063102-2539cc00-f1cb-11e8-97ea-5bceabf10434.jpg)
![cond1](https://user-images.githubusercontent.com/22091149/49063108-2bc84380-f1cb-11e8-86fb-794bd61c0556.jpg)
![dmlog1](https://user-images.githubusercontent.com/22091149/49063123-3aaef600-f1cb-11e8-8fbd-08206fa9f89e.jpg)
![dmlog2](https://user-images.githubusercontent.com/22091149/49063130-400c4080-f1cb-11e8-97d9-6de21e7a14ab.jpg)

**Comment by SAM26K** [comment]
All those screenshots are from the master pi on 192.168.1.17 and has the A/D shield mounted on it.  The pi controlling the relays is on 192.168.51.

**Comment by SAM26K** [comment]
Screenshots from remote relay pi. 8 Channel relay board controlled with GPIO  has daemon errors but works ok from master pi using manual controls:
![outpi2](https://user-images.githubusercontent.com/22091149/49063839-90849d80-f1cd-11e8-8f84-fa5d0324a757.jpg)

**Comment by kizniche** [comment]
Do all conditionals stop working at the same time?

**Comment by kizniche** [comment]
Do you have the ability to test the 7.0 branch for this issue? I'm hesitant to put in a lot of time diagnosing an issue that may have already been fixed in the next release. [This comment](https://github.com/kizniche/Mycodo/issues/562#issuecomment-441150035) contains the commands to get the 7.0 branch running.

**Comment by SAM26K** [comment]
Will try that.  FYI the only way to get the conditional working again is to
delete the comparison parameter and re-enter it again.
Does the 7.0 branch have the ADS1256 fixes?
Thanks

**Comment by kizniche** [comment]
Yes, it should be working.

**Comment by SAM26K** [comment]
Also noticed I have a lot more input choices with the ADS1256 A/D pcb.  Only 8 ch AD is connected to pi.  Have two inputs in the data page and 2 measurement displays on dash.  What are these extra inputs? TIA
![volt_dash3](https://user-images.githubusercontent.com/22091149/49335112-91397d00-f59b-11e8-8480-9f1acd6b48cb.jpg)

**Comment by SAM26K** [comment]
Data input control settings for 2nd  analog input:
![volt_dara3](https://user-images.githubusercontent.com/22091149/49335164-0eb1bd00-f59d-11e8-90d2-cfc72349b51f.jpg)

**Comment by kizniche** [comment]
Fixed the decimal issue in the last commit (above).

You'll have to explain in more detail what the ADC issue is, as it looks to me to be working properly. You have two ADC Inputs, each with 8 channels. That means each input (8 channels each) will appear as selection options (totaling 16 channels).

**Comment by kizniche** [comment]
It looks from the screenshot you are only enabling 1 of the 8 channels for Input ACh1. This indicated you only have one thing connected to this ADC (on channel 2). I can't see the other Input to determine how you have configured it, but I suspect your second physical ADC only has one input connected as well and you have only one channel enabled.

I'm not sure why you don't have both connections connected to just one ADC chip, channel 1 and channel 2. It seems like a waste of the second ADC when you could have both inputs connected to just one ADC.

**Comment by kizniche** [comment]
I also just noticed another Input in the screenshot, with a third ADC chip. I am now suspecting you are improperly configuring the ADC. You should only have one Input per ADC chip.

**Comment by kizniche** [comment]
I need to add some sanity-checking to prevent more than one Input with the same I2C address and bus from being activated. This will prevent such issues with misunderstanding how to configure an ADC.

**Comment by SAM26K** [comment]
Only using a single ADS1256 Waveshare PCB.  The new 7.0 controls seem to
generate multiple outputs

**Comment by SAM26K** [comment]
What about the resolution?       Dash > measurement shows 0 decimal places.

**Comment by kizniche** [comment]
> The new 7.0 controls seem to generate multiple outputs

Do you mean inputs? There will be as many channels as the chip will allow. for the ADS1256, there will be 8 channels. When you add the input, all 8 are enabled and you can disable any you don't need a measurement from.

> What about the resolution? Dash > measurement shows 0 decimal places.

I already pushed a fix, so you will need to grab the latest code with:

```cd ~/Mycodo && sudo git fetch && sudo git pull && sudo /bin/bash ~/Mycodo/mycodo/scripts/upgrade_commands.sh initialize && sudo service mycodoflask restart```

**Comment by SAM26K** [comment]
Decimals working ok but now if more than 2 channels from the ADS1256 is enabled, the all go to just above zero volts out.  Also notice that the input selection increases with every input enabled.
![volt_dash5](https://user-images.githubusercontent.com/22091149/49352205-aed61780-f66b-11e8-9e7e-64a1fffa5b7d.jpg)
![volt_dash4](https://user-images.githubusercontent.com/22091149/49352207-af6eae00-f66b-11e8-9802-88f09dd3908f.jpg)
![volt_live3](https://user-images.githubusercontent.com/22091149/49352208-af6eae00-f66b-11e8-8d6c-d00b62a1afc0.jpg)
![volt_data3](https://user-images.githubusercontent.com/22091149/49352209-af6eae00-f66b-11e8-8c5a-6a41144998fa.jpg)

**Comment by SAM26K** [comment]
![damian4](https://user-images.githubusercontent.com/22091149/49352280-01afcf00-f66c-11e8-8ccb-9fe3ea8239ec.jpg)

**Comment by kizniche** [comment]
First, you must start using the input properly before we can assess if there is an issue. Delete all inputs except for one.

**Comment by SAM26K** [comment]
All inputs deleted and created a single ch0 data input,  now getting a 500 server error, should I re-install?
![dash6](https://user-images.githubusercontent.com/22091149/49353656-b7325080-f673-11e8-9419-ca69bc64fa9f.jpg)

**Comment by SAM26K** [comment]
How do I use the new inputs properly?

**Comment by kizniche** [comment]
I just pushed a fix for the 500 error. You'll have to do the series of commands from above with git fetch/pull to update.

You should only have one ADC input and enable/disable the channels you want to use in the "Measurements Enabled" multiselect box.

**Comment by SAM26K** [comment]
Ok back to working again but still confused about the new config of AD, Now one instance of the input has 8 channels each instead of 1 channel?
I think only the first channel is accurate using that method.  ie if I   pick channel 2 of instance one it reads  near a rail and know its not accurate, If I create a new instance of the ADS1256 with the next channel, that channel is accurate too,

**Comment by SAM26K** [comment]
Where is the  ""Measurements Enabled" multiselect box"?

**Comment by SAM26K** [comment]
Ok, I think I'm getting it now.  Will take a while to rebuild all the controls but so far seeing reasonable numbers for the 6 channels I have sensors connected to.

**Comment by SAM26K** [comment]
Hit a 500 error when adding outputs to dashboard:

Error 500: Internal Server Error
Something bad happened but it's probably not your fault. Letting the developers know about these issues is crucial to supporting Mycodo. Please submit a new issue on GitHub with the following error traceback (copy the entire traceback):

Error (Full Traceback):

Traceback (most recent call last):
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask_login/utils.py", line 261, in decorated_view
    return func(*args, **kwargs)
  File "/home/pi/Mycodo/mycodo/mycodo_flask/routes_page.py", line 685, in page_dashboard
    y_axes=y_axes)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/templating.py", line 135, in render_template
    context, ctx.app)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/flask/templating.py", line 117, in _render
    rv = template.render(context)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/home/pi/Mycodo/mycodo/mycodo_flask/templates/pages/dashboard.html", line 3, in top-level template code
    {% set help_page = ["dashboard", _('Dashboard')] %}
  File "/home/pi/Mycodo/mycodo/mycodo_flask/templates/layout.html", line 259, in top-level template code
    {%- block body %}{% endblock -%}
  File "/home/pi/Mycodo/mycodo/mycodo_flask/templates/pages/dashboard.html", line 63, in block "body"
    {% include 'pages/dashboard_options/mod_graph.html' %}
  File "/home/pi/Mycodo/mycodo/mycodo_flask/templates/pages/dashboard_options/mod_graph.html", line 190, in top-level template code
    {{' (' + dict_units[colors_graph[each_dashboard.unique_id][n]['unit']]['name']}})</label>
  File "/home/pi/Mycodo/env/lib/python3.5/site-packages/jinja2/environment.py", line 411, in getitem
    return obj[argument]
jinja2.exceptions.UndefinedError: collections.OrderedDict object has no element Undefined

**Comment by kizniche** [comment]
I just pushed a fix for the graph issue adding outputs.

**Comment by SAM26K** [comment]
Thanks for the update. Seems to work ok so far but still rebuilding all the controls.  Not updating the slave pi with the relay GPIO PCB attached.  Seems to work ok with 6.45 so far.
Noticed that the dashboard 'measurement' display height parameter seems to no longer have any effect with 7.0.
Thanks again.

**Comment by SAM26K** [comment]
Seems to be a problem with timed output.  If I set the conditional to turn on for 2 mins and then ignore for 5 mins the output turns on but never turns off. Manual controls ok.  e.g. 
![master_f5](https://user-images.githubusercontent.com/22091149/49640314-2389b880-f9c2-11e8-95e5-da15f3b36a2c.jpg)

**Comment by kizniche** [comment]
It seems the conditional refractory period wasn't operating as expected. I just pushed a fix for that, however I've also pushed a commit recently that changed the database schema (added a start offset to PID and conditional controllers), so your current database will need to be regenerated if you want to use the latest code.

If you want to just make the very minor change yourself, all you need o do is change the following line:

https://github.com/kizniche/Mycodo/blob/42244940998a887603836e4d961642b6b9ce5622/mycodo/controller_conditional.py#L275

To this:

```python
        if self.refractory_period:
```

Save, then restart the backend.

**Comment by SAM26K** [comment]
Thanks for the update.  Not updating dB yet as not using PID functions yet.  Focused on verifying if conditionals stop working in 7.0 or not.  Do you think the other changes could affect the conditionals reliability problem?
Thanks again for the code fix.

**Comment by SAM26K** [comment]
The latest issue was that the conditional function stopped working after a
while and had to rewrite the conditional parameter and resave the control
to get it to respond again.
Would you prefer I start a new issue with new issues related to conditional
functions?

**Comment by kizniche** [comment]
What I meant was I understood what the issue was, but I was not able to replicate it. Don't open a new issue unless you can replicate it in the 7.0 branch.

**Comment by SAM26K** [comment]
Everything seems to be working so far except for function > conditional > refractory period.  If I set it to 300 secs it only runs for 2.5 minutes before it activates again.

**Comment by kizniche** [comment]
I just released 7.0.0. Can you install this release fresh and test if this still occurs? I wasn't able to replicate it. When I set the refractory period to 300, it waits 5 minutes from execution to execute again (if conditional remains true).

**Comment by SAM26K** [comment]
Is there a way to use the web UI "upgrade" function or how should I upgrade?  It tells me I'm already up to date even though my version is edited.  Is there a way to keep all the controls?
TIA

**Comment by SAM26K** [comment]
Were you able to confirm the dashboard measurement display's height
parameter doesn't seem to have any effect?

**Comment by kizniche** [comment]
Are you suggesting only Measurements on the dashboard exhibit this behavior? To my understanding, the new version of bootstrap (v4) doesn't allow different height divs on the same row. Therefore, you should see this represent as all elements on the same row will be the height of the tallest element.

**Comment by SAM26K** [comment]
Ok, you're right.  No problem there.
Thanks again.

**Comment by SAM26K** [comment]
Will need more time to confirm operation to close the original issue though.

**Comment by kizniche** [comment]
> Ok, you're right. No problem there.

I spent a good while trying to solve this issue, but it seems fairly difficult. I may return to it soon. Having different heights in the same row is something I liked.

**Comment by SAM26K** [comment]
Is there any to reduce font size below 1?  I'm using 1 for all font sizes
and could use 1/4 that size with a 1080 monitor resolution.

**Comment by kizniche** [comment]
Font size of what? You can zoom out in your browser.

**Comment by SAM26K** [comment]
Was referring to the font size parameter in various display widgets.  Using
1 for everything but maybe a completely different font could produce
smaller screen fonts? Know about the browser zoom function but that is not
the best solution.
No big deal just a suggestion when you get time.
Thanks again.

**Comment by kizniche** [comment]
You can set the font size options to less than 1.

**Comment by SAM26K** [comment]
Did a fresh install from the released  7.0 and rebuilt the web UI controls. 
The problem now is that when I create  > function > conditional,  it no longer lists the id parameter in curly brackets. Instead, it generates a generic text that is the same for all new function conditional controls.
ie its not clear what  {} is needed for the comparison test:
![cond1](https://user-images.githubusercontent.com/22091149/49698947-9ee49900-fb7f-11e8-80b3-9cf5293e0736.jpg)

**Comment by kizniche** [comment]
I make a quick update to make the conditional system a little more powerful, now that full, long Python code can be used. It's at the expense of usability to the inexperienced, but I'd like to get your feedback on how the documentation can be improved. The screenshot, below, has the ID circled that you need to use in the code.

![49698947-9ee49900-fb7f-11e8-80b3-9cf5293e0736](https://user-images.githubusercontent.com/838427/49699027-d3fde500-fb99-11e8-8818-56318f63dd32.jpg)

**Comment by kizniche** [comment]
I should change the [ID] in that disabled input/button to {ID} to make it a little more suggestive that is the correct ID.

**Comment by SAM26K** [comment]
Ok thanks for the update.  Do I need to push that change manually or how should  I do that?

**Comment by kizniche** [comment]
Since 7.0.0 is released, you'll get the updates as you normally would when I release 7.1.1, which should be sometime today.

**Comment by SAM26K** [comment]
OK, thanks again.

**Comment by kizniche** [comment]
I also just added a sanity check for the actual code. The IDs will be attempted to be replaced and the code executed before it's allowed to be saved (screenshot, below). If there's an error, a large detailed error will appear and the code will be prevented from being saved.

![screenshot_2018-12-09 delta - function - mycodo 7 0 0](https://user-images.githubusercontent.com/838427/49700431-70c97e00-fbac-11e8-9ebb-3af7ed610d21.png)

**Comment by SAM26K** [comment]
Hi Kyle,
With fresh install using released 7.0.0 conditional controls no longer work at all.  With 7.0.1 I get an error similair to above.  The ID was copy and pasted from the condition tag.
![dash1](https://user-images.githubusercontent.com/22091149/49709740-bc9b1800-fbea-11e8-8cb0-167e28ddab9a.jpg)
![function4](https://user-images.githubusercontent.com/22091149/49709741-bc9b1800-fbea-11e8-8ce4-ad5ed81b57a7.jpg)
![functionerr1](https://user-images.githubusercontent.com/22091149/49709742-bc9b1800-fbea-11e8-937f-11c9a11323ec.jpg)

**Comment by kizniche** [comment]
Back at my computer. For syntactically-correct Python code, our conditional statement should be:

```python
if {66493fa3} > 1.8:
    print(1)
```

**Comment by SAM26K** [comment]
Thanks, I got it working now.  Will run for a few days to confirm it keeps working.
Thanks.

**Comment by SAM26K** [comment]
It still seems to have a timing problem.  I'm programming it to turn on a relay for 2 minutes then ignore for five minutes.  Instead of five minutes, it gets triggered again 2.5 minutes later instead of 5 minutes.  The weird thing is the graph shows five minute intervals between relay activations.  My stopwatch shows 2.5 minutes between activations.

**Comment by SAM26K** [comment]
It seems I was measuring wrong.  If I time it from the start of the trigger to the start of the following trigger it is 5 minutes.

**Comment by kizniche** [comment]
While setting up conditionals for one of my production systems today with the new 7.0, I found what can be an issue for particular setups and wanted to let you know about my findings.

The Max Age settings for a measurement determines how long ago to look in the database for a measurement. If it finds a measurement, it will replace the {ID} with the actual measurement before running the Python code. However, if it doesn't find a measurement, it will replace {ID} with "None". Therefore, if you don't want the code to throw an error (for trying to compare None to a numerical value, such as "None < 23", then you will need to add a test for None-ness before the actual conditional comparing it to a value. This can easily be accomplished a number of ways, but here are a few examples:

Example statement:

```python
if {asdf1234} < 24:
    print(1)
```

This statement will cause an error if {asdf1234} is replaced with None. However, if we add a check for None before the comparison to 24, we can avoid the error:

```python
if {asdf1234} is not None and {asdf1234} < 24:
    print(1)
```

or

```python
if {asdf1234} is not None:
    if {asdf1234} < 24:
        print(1)
```

I duplicated the check during saving of the conditional for properly-formatted code to include replacing with None (in addition to replacing with a numerical value). Now, if the code isn't formatted properly if a value has a None value, then the error will appear, like below:

![screenshot_2018-12-10 micromyco - function - mycodo 7 0 1](https://user-images.githubusercontent.com/838427/49763783-a0a37f00-fc9b-11e8-8e0e-4b154d2e4522.png)

**Comment by kizniche** [comment]
However, I find it useful to include a conditional that makes sure measurements are being acquired, and if they aren't (returns None), then it will email me (or turn off the system and email me).

```python
if {asdf1234} is None:
    print(1)
```

**Comment by SAM26K** [comment]
Thanks for the heads up.

**Comment by SAM26K** [comment]
If I run "Test all actions" it cycles the relay but no email shows up.

**Comment by kizniche** [comment]
I don't understand why you have "print(2)" there. Did you set up the email credentials in the configuration and test sending an email from there?

**Comment by SAM26K** [comment]
Got test email working.  Had to reduce security to allow Mycodo to logon to
gmail account. Now test all functions works with no errors.  But the
conditional never triggers with that extra print line. Isn't that the
format for sending an email (print(2)) to execute the 2nd action? Just
testing.

**Comment by SAM26K** [comment]
Ok, that works.  Thanks.  Yes, it would be nice to have separate controls
for more complex conditions.
Thanks again.

**Comment by SAM26K** [comment]
Hi Kyle,
Conditional reliability has been really good the last few days and no longer seeing the original problem of stopping.  Only problem I'm having now is with the condition parameter.  If I set it to check one channel per control it works fine.  But if I check more than one channel it triggers every time even though the condition is false.  Am I doing something wrong?
![email1](https://user-images.githubusercontent.com/22091149/50150194-172f1680-0272-11e9-8b67-53c9cf910610.jpg)
![email2](https://user-images.githubusercontent.com/22091149/50150225-23b36f00-0272-11e9-93e9-bd874057de85.jpg)
TIA.

**Comment by kizniche** [comment]
Your conditional is currently:
If 1.7 (True) or 1.5 (True) or 1.7 (True) or 1.4 (True) or None (False), then trigger actions.
It appears to be working as you coded it. Without telling me what you're trying to accomplish, I won't be able to tell you how it differs from what it should be.

**Comment by SAM26K** [comment]
Was trying to test for a dead (null) channel.  Thought that it would be
false if all valid and true if any were null?

**Comment by kizniche** [comment]
To test if one channel isn't returning a measurement:

```python
if {ID} is None:
    print(1)
```

To test if any aren't returning a measurement:

```python
if None in [{ID1}, {ID2}, {ID3}, {ID4}]:
    print(1)
```

**Comment by SAM26K** [comment]
Thanks for the correction.  Looks like I need to read up more on python syntax.
I was wondering about this other problem I'm seeing.  I set up a conditional control to turn on a solenoide which opens a drip line, wait 5 secs and then turn on a water pump.  The sequence seems to work ok and timing is good but when it shows up on the graph there is less than 1 sec spacing between the switching action even though its really 5 secs.  The cursor is offset for some reason too but not concerned with that.:
![cond11](https://user-images.githubusercontent.com/22091149/50200731-0b902e00-030c-11e9-84f9-907167ee7bf9.jpg)
![dash11](https://user-images.githubusercontent.com/22091149/50200736-134fd280-030c-11e9-9fa7-e78b6ff46086.jpg)
![dash12](https://user-images.githubusercontent.com/22091149/50200746-1cd93a80-030c-11e9-8772-d970c2389ef2.jpg)
TIA

**Comment by kizniche** [comment]
It looks like 5 seconds to me. Without you showing both timestamps we can't calculate the actual time, but with 15 seconds between demarcations of your x axis, that appears to be separated by 5 seconds.

**Comment by kizniche** [comment]
Oh it appears they are overlapping. I'll do some testing.

**Comment by kizniche** [comment]
In the future, create a new issue if it is not relevant to the current issue. However, I will respond to your last comments, but please try to stay on topic.

I haven't been able to replicate your conditional timing/graph issue. This may be an issue with interpreting the graph. I would need to see a more zoomed-in graph and it render it properly.

As for the new sensor, all you would need to do is create a new input and import it through the Input menu of the Configuration menu. There are some changes slated for the 7.0.2 release that enable setting the timestamp from the Input module, so in order to use the module below, you'll either need to wait for the 7.0.2 release or [upgrade to the master branch](https://github.com/kizniche/Mycodo/wiki/Updating-to-Github-Master-Branch).

New Input Module ```Input Module - TH16.py```:

```python
# coding=utf-8
import datetime
import json
import logging

from mycodo.databases.models import DeviceMeasurements
from mycodo.inputs.base_input import AbstractInput
from mycodo.inputs.sensorutils import calculate_dewpoint
from mycodo.inputs.sensorutils import calculate_vapor_pressure_deficit
from mycodo.inputs.sensorutils import convert_from_x_to_y_unit
from mycodo.utils.database import db_retrieve_table_daemon

# Measurements
measurements_dict = {
    0: {
        'measurement': 'temperature',
        'unit': 'C'
    },
    1: {
        'measurement': 'humidity',
        'unit': 'percent'
    },
    2: {
        'measurement': 'dewpoint',
        'unit': 'C'
    },
    3: {
        'measurement': 'vapor_pressure_deficit',
        'unit': 'Pa'
    }
}

# Input information
INPUT_INFORMATION = {
    'input_name_unique': 'TH16',
    'input_manufacturer': 'Sonoff',
    'input_name': 'TH16',
    'measurements_name': 'Humidity/Temperature',
    'measurements_dict': measurements_dict,

    'options_enabled': [
        'measurements_select',
        'period',
        'pre_output'
    ]
}


class InputModule(AbstractInput):
    def __init__(self, input_dev, testing=False):
        super(InputModule, self).__init__()
        self.logger = logging.getLogger("mycodo.inputs.th16")
        self._measurements = None

        if not testing:
            self.logger = logging.getLogger(
                "mycodo.th16_{id}".format(id=input_dev.unique_id.split('-')[0]))

            self.device_measurements = db_retrieve_table_daemon(
                DeviceMeasurements).filter(
                DeviceMeasurements.device_id == input_dev.unique_id)

    def get_measurement(self):
        return_dict = measurements_dict.copy()

        #
        # Do stuff here to get str_json, then comment or delete the following
        # lines used for debugging.
        #

        str_json = '{"StatusSNS":{"Time":"2018-12-21T13:06:46","AM2301":{"Temperature":69.1,"Humidity":54.1},"TempUnit":"F"}}'

        dict_data = json.loads(str_json)

        # Convert string to datetime object
        datetime_timestmp = datetime.datetime.strptime(
            dict_data['StatusSNS']['Time'], '%Y-%m-%dT%H:%M:%S')

        # Convert temperature to SI unit Celsius
        if self.is_enabled(0):
            return_dict[0]['value'] = convert_from_x_to_y_unit(
                'F', 'C', dict_data['StatusSNS']['AM2301']['Temperature'])
            return_dict[0]['timestamp'] = datetime_timestmp

        if self.is_enabled(1):
            return_dict[1]['value'] = dict_data['StatusSNS']['AM2301']['Humidity']
            return_dict[1]['timestamp'] = datetime_timestmp

        if (self.is_enabled(2) and
                self.is_enabled(0) and
                self.is_enabled(1)):
            return_dict[2]['value'] = calculate_dewpoint(
                return_dict[0]['value'], return_dict[1]['value'])
            return_dict[2]['timestamp'] = datetime_timestmp

        if (self.is_enabled(3) and
                self.is_enabled(0) and
                self.is_enabled(1)):
            return_dict[3]['value'] = calculate_vapor_pressure_deficit(
                return_dict[0]['value'], return_dict[1]['value'])
            return_dict[3]['timestamp'] = datetime_timestmp

        return return_dict
```

**Comment by kizniche** [comment]
v7.0.2 was just released

**Comment by SAM26K** [comment]
Thanks a lot for that new TH16 input.  It seems to be working fine except for a minor issue with the dashboard measurement control displaying deg C instead of deg F (input has convert to F selected).  Should I open a new issue?  Thanks again.
![dashth](https://user-images.githubusercontent.com/22091149/50370073-dd515f00-0554-11e9-8a03-a382ce1eb2e1.jpg)

**Comment by kizniche** [comment]
Yes, please create a new issue. It should be an easy fix, but I don't think I'll get to it today and this closed issue will get lost/forgotten.
